### ISSUE TITLE
A new module system change/regression?

### ISSUE THREAD

**Issue by WolframKahl** [agda/agda#1815]
While I had been waiting for my problem with issue #1692 to go away, something has changed in the module system ─ the attached module typechecks neither in the current development version (9081bb140b8e880c45c2cb720bf8ee0f886b4a89), nor in the 2.4.2.6 branch (fe788f8da47e2662ad6904f51c3d5e05297b01ca). However, it, and the original module it has been pruned from, typecheck with 2.4.2.3 and with 6ede17d362450d520076e53916277eeba311dc28; indeed, this typechecked for years without changes.

[NewRecordModuleError.agda.txt](https://github.com/agda/agda/files/116506/NewRecordModuleError.agda.txt)

I find nothing in that direction in the release notes, neither in master, nor in the 2.4.2.6 CHANGELOG, so I guess that this may not be an intentional change.

I haven't had time to minimise it, so I just produced a slice in a single file requiring the standard library.
(Each of my installations includes the corresponding standard library; this uses only Relation.Binary, so the version should not really matter.)

Recent versions:

```
/var/tmp/kahl/svn/A/RATH-Agda/trunk/NewRecordModuleError.agda:78,35-36
(LowerBound₂ 𝒫 R S) !=<
(Poset (_p₁_124 𝒫 g) (_p₂_125 𝒫 g) (_p₃_126 𝒫 g)) of type
(Set (k₂ ⊍ j))
when checking that the expression m has type
Poset (_p₁_124 𝒫 g) (_p₂_125 𝒫 g) (_p₃_126 𝒫 g)
```

Older versions:

```
Finished NewRecordModuleError.
```

**Comment by asr** [comment]
I could type-check the example with Agda 2.4.2.5.

**Comment by WolframKahl** [comment]
On Thu, Feb 04, 2016 at 04:16:52AM -0800, Andrés Sicard-Ramírez wrote:

> I could type-check the example with Agda 2.4.2.5.

This supports my impression that this must be a rather recent change.
(Agda 2.4.2.5 has other differences (wr.t issue #1692) that make it currently unusable for me.)

**Comment by andreasabel** [comment]
It looks like I introduced this regression by https://github.com/agda/agda/commit/9c64919edb6d0b98b6b89ebce4f5f66c0fc0aa91, fixing issue #1759.

Somehow the assumption is used that in a parametrized module, all definitions share the same module telescope.  However, this is not true for record modules, where e.g. fields have hidden parameters, but other definitions keep the module parameters as written, i.e., they could be visible, like the parameter `P` of module `LowerBounds`.
When the module `IsLowerBound2 proof` in record `LowerBound2` is formed, somehow the currently active module telescope (where `P` is visible) overrides the telescope in the types of `bound1` and `bound2`, where `P` is correctly hidden.

**Comment by andreasabel** [comment]
Here is a short test case:

``` agda
module Issue1815 where

module _ (A : Set) where

  record R1 : Set where
    field f1 : A

  record R2 : Set where
    field f2 : R1
    open R1 f2 public

test1 : ∀ A (r : R1 A) → A
test1 A r = R1.f1 r

worksShouldFail : ∀ A (r : R2 A) → A
worksShouldFail A r = R2.f1 A r

test2 : ∀ A (r : R2 A) → A
test2 A r = R2.f1 r

-- Error:
-- (R2 .A) !=< Set of type Set
-- when checking that the expression r has type Set
```

**Comment by andreasabel** [comment]
Ulf suggests: I would expect

``` agda
  module M Γ where
    record R Δ
```

to yield the same types for functions and projections as

``` agda
  module M where
    record R Γ Δ
```

This can surely be implemented, but to make this stringent, we have to rethink what it means to abstract a telescope over definitions.

Think food 1:

``` agda
module _ (S : Set) where

-- Variant 1: Parametrized module definition

module M (A : Set) where
  record R : Set where
    field      f : S
    postulate  P : Set

-- Variant 2: Abstracting parameter later.

module N' where
  record R : Set where
    field     f : S
    postulate P : Set
module N (A : Set) = N'

-- Compositionality would require them to behave the same.

-- They don't, for field f.

-- Nf = N.R.f {S}  -- fails
Mf = M.R.f {S}  -- passes

-- They do, for postulate P, after fixing #1759.

NP = N.R.P S
MP = M.R.P S  -- fails with 2.4.2.5, passes with 2.4.2.6
```

Think food 2:  Should not we expect 

``` agda
module M' (A : Set) = M A
```

to create `M'` as an identical copy of `M`?

``` agda
M'f = M'.R.f S  -- passes, while it fails for M (needs hidden {S}).
```

**Comment by UlfNorell** [comment]
I do think ideally that the `N` and `M` modules should behave the same in example 1, and that `M'` and `M` should be the same in example 2. As you observe, this would require rethinking abstraction and application for definitions. My suggestion for this is to have abstraction over a record/data type be parameter insertion. A complication in that case is that record modules cannot be treated as regular modules since the parameter telescope must be made hidden.

I do think it makes sense to not require a user to understand precisely how record modules are desugared in order to understand what happens when you put one in a parameterised module.

**Comment by WolframKahl** [comment]
(I have adapted my code to work around this issue by defining (private) modules and accessing the problematic fields via those. So far I have found only two places where that was necessary. Those modules are defined using the same actual parameters in both versions, so my code should now be unaffected by this issue. If this changes against 2.4.2.5, I think it should be documented, and would in principle not be the kind of change I expect from a maintenance release, but since it is easy to work around and probably does not affect that many users, it could be OK.)

**Comment by andreasabel** [comment]
Ulf, your expressed expectation is not what is being implemented, see `test/Succeed/RecordInParModule.agda`.  Here is a variant of this test case:

``` agda
module RecordInParModule (X : Set) where

module Works where

  module M where
    record R (x : X) : Set where

  module E {x : X} (r : M.R x) where
    open module M' = M.R r

module WorksInOldSemanticsButFailsInNewSemantics where

  module M (x : X) where
    record R : Set where

  module E {x : X} (r : M.R x) where
    open module M' = M.R x r
```

If `x` is a parameter to the record type `R` (see `module Works`) then module `M.R` only expects `r` visibly.  But if we pull the parameter out to the parent `module M`, then module `M.R` expects both `x` and `r` visibly.  
Currently, the `sectionTel`s differ for the two variants of `R`.  Thus, the new fix for #1759, which just adapts `freeVarsToApply`, does not cover the latest case of issue #1759, where `postulate P` is also inside a record.

**Comment by andreasabel** [comment]
This regression is defeated, like Napoleon in Waterloo, but we can continue to discuss the semantics of parametrized modules and records here.
